Fixed popup misalignment with map during pan easing

bicubic opened the pull request:
Popup DOM transforms are deferred until `painter.render()` except when
Popup is being instantiated.

Popup offset integer rounding is suppressed during easing animations to
maintain sub-pixel alignment. The offset is rounded off during the final
tick of an easing animation to ensure DOM is pixel perfect on
completion.

DOM.setTransform takes new optional `immediate` parameter. If false, the
transform will be deferred until next canvas render. This change is
backwards compatible.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [x] briefly describe the changes in this PR
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page

jfirebaugh commented on src/ui/popup.js:
```diff
@@ -272,7 +272,12 @@ class Popup extends Evented {
+        let offsetedPos = this._pos.add(offset[anchor]);
+        //snap to interger offset if this is last tick of an easing
+        //animation, or if this event didn't come from easing loop
+        if (ev && (ev.finalEasingTick === true || ev.easing !== true)) {
```
[`Marker` has similar logic](https://github.com/mapbox/mapbox-gl-js/blob/213d18a501fd505a846448599ca57d56a9469967/src/ui/marker.js#L153-L153), but the condition is based on checking the event type (`move` vs `moveend`). Will that work here too, instead of adding new properties to the event object?

bicubic replied:
Thanks @jfirebaugh . It should work, I will look at folding that in after @mourner posts his feedback.

bicubic replied:
@jfirebaugh @mourner do you guys have any further input or are you happy to merge once I look at switching to `moveend`

jfirebaugh commented:
Thanks for tracking down the cause here and submitting a PR @bicubic! I have one question about the rounding part, and I think @mourner wants to look at the transform part.

bicubic commented:
Alright, @mourner as per your feedback, transform management state has been moved to `Map`. 
`util.pushTransform` and `DOM.setTransform` now take a new optional map argument. As before, both functions are backwards compatible and using the old call signature will work the same as before. 

@jfirebaugh I have looked at your suggestion, unfortunately `moveend` is never received by `popup._update()` in this scenario and I don't have a good enough view of your codebase to understand why Marker receives that event. My guess is it's catering to a different use case and doesn't deal with pan easing (comments seem to refer to zooming there). I'm inclined to leave my workaround in place unless you feel strongly against it and are able to help me augment pan easing to fire `moveend` on the last tick.

bicubic commented:
Alright, @mourner as per your feedback, transform management state has been moved to `Map`. 
I have removed the previous util functions and reverted DOM.setTransform to the way it was in develop. Instead, callers must explicitly call the new `Map.deferTransform` method, e.g.:
```
if (instantiating === true) {
    DOM.setTransform(this._container, transform); //exactly the same as before, applied immediately
} else {
    this._map.deferTransform(this._container, transform); //coalesced, deferred until next render tick
}
```

The reason DOM.setTransform must continue to exist is because the transform must be set immediately when the overlay dom is created, deferring it causes the element to be out of position for one frame.

@jfirebaugh I have looked at your suggestion, unfortunately `moveend` is never received by `popup._update()` in this scenario and I don't have a good enough view of your codebase to understand why Marker receives that event. My guess is it's catering to a different use case and doesn't deal with pan easing (comments seem to refer to zooming there). I'm inclined to leave my workaround in place unless you feel strongly against it and are able to help me augment pan easing to fire `moveend` on the last tick.

jfirebaugh commented:
Thanks @bicubic, however this still doesn't feel like the right approach to me. The following issues give me pause:

* Why doesn't `moveend` work? I can't think of any reason why it wouldn't. This needs investigation.
* I'm not happy with introducing additional state to the map in the form of "deferred transitions". That seems like an approach that could cause race conditions or other unexpected behavior. Are there any other approaches you can think of?

bicubic commented:
Thanks @jfirebaugh 

> Why doesn't moveend work? 
I'll leave that up to you guys to figure out. I don't have the time to trace your eventing architecture nor do I know the intents of the original author. 

>  That seems like an approach that could cause race conditions or other unexpected behavior. Are there any other approaches you can think of?

Could look at having an event fired by the map immediately after painter.render, an event that's exclusively used to signal markers and popups to execute CSS state chaneges. However, I can't imagine a scenario where that would have any objective advantages over having the map store transition state. That solution would  be logically equivalent to the current one, but with state ownership being held by the popups instead of map. We can look into that if you can show a concrete case where the current approach would exhibit race conditions or unexpected behaviour.

mourner commented:
Superceded by #5743
